stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt

```diff
@@ -0,0 +1,797 @@
+        _user.value = me
+        _mutedUsers.value = me.mutes
+        _channelMutes.value = me.channelMutes
+        _totalUnreadCount.value = me.totalUnreadCount
+        _channelUnreadCount.value = me.unreadChannels
+        _banned.value = me.banned
+    }
+
+    private infix fun UserId.mustBe(currentUserId: UserId?) {
+        if (this != currentUserId) {
+            throw InputMismatchException(
+                "received connect event for user with id $this while for user configured " +
+                    "has id $currentUserId. Looks like there's a problem in the user set"
+            )
+        }
+    }
+
+    companion object {
+        val EMPTY_DISPOSABLE = object : Disposable {
+            override val isDisposed: Boolean = true
+            override fun dispose() {}
+        }
+    }
+}
+
+private typealias UserId = String
+
+private class BatchEvent(
+    val sortedEvents: List<ChatEvent>,
+    val isFromHistorySync: Boolean,
+) {
+    val size: Int get() = sortedEvents.size
+    val isFromSocketConnection get() = !isFromHistorySync
+}
+
+private class SocketEventCollector(
+    private val scope: CoroutineScope,
+    private val fireEvent: suspend (BatchEvent) -> Unit
+) {
+    private val mutex = Mutex()
+    private val postponed = arrayListOf<ChatEvent>()
+    private val timeoutJob = AtomicReference<Job>()
+
+    suspend fun add(event: ChatEvent): Boolean {
+        if (event is UserStartWatchingEvent || event is UserStopWatchingEvent) {
+            StreamLog.d(TAG) { "[add] event.type: ${event.type}" }
+            mutex.withLock {
+                return postponed.add(event).also {
+                    scheduleTimeout()
+                }
+            }
+        }
+        return false
+    }
+
+    private fun scheduleTimeout() {
+        timeoutJob.get()?.cancel()
+        timeoutJob.set(
+            scope.launch {
+                delay(TIMEOUT)
+                StreamLog.i(TAG) { "[scheduleTimeout] timeout is triggered" }
+                doFire()
+                timeoutJob.set(null)
+            }
+        )
+    }
+
+    suspend fun fireBatchEvent() {
+        StreamLog.d(TAG) { "[fireBatchEvent] no args" }
+        timeoutJob.get()?.cancel()
+        timeoutJob.set(null)
+        doFire()
+    }
+
+    private suspend fun doFire() {
+        mutex.withLock {
+            if (postponed.isEmpty()) {
+                StreamLog.v(TAG) { "[doFire] rejected (postponed is empty)" }
+                return
+            }
+            StreamLog.v(TAG) { "[doFire] postponed.size: ${postponed.size}" }
+            val sortedEvents = postponed.toList().sortedBy { it.createdAt }
+            postponed.clear()
+            fireEvent(
+                BatchEvent(sortedEvents, isFromHistorySync = false)
+            )
+        }
+    }
+
+    private companion object {
+        private const val TAG = "Chat:EventCollector"
+        private const val TIMEOUT = 300L
```

Comment:
Just out of curiosity - why 300?

Reply:
Based on max gap of 200ms I've seen in logs between 2 events